types/coin.go

```diff
@@ -508,15 +513,49 @@ func ParseCoin(coinStr string) (coin Coin, err error) {
 
 	denomStr, amountStr := matches[2], matches[1]
 
-	amount, ok := NewIntFromString(amountStr)
-	if !ok {
-		return Coin{}, fmt.Errorf("failed to parse coin amount: %s", amountStr)
-	}
-
 	if denomStr != strings.ToLower(denomStr) {
 		return Coin{}, fmt.Errorf("denom cannot contain upper case characters: %s", denomStr)
 	}
 
+	// test if the amount is specified using scientific notation
+	scinMatches := reScinAmt.FindStringSubmatch(amountStr)
+
+	ok := false
+	var amount Int;
+
+	if len(scinMatches) == 4 {
+
+		// Parse scientific notation such as 1.23e+12stake
+		var f float64
+		var dec int
+		
+		if f, err = strconv.ParseFloat(scinMatches[1], 64); err != nil {
+			return
+		}
+		
+		if dec, err = strconv.Atoi(scinMatches[3]); err != nil {
+			return 
```

Comment:
we're trying to name _all_ explicit return variables within the code base now (aka this should be `return Coins{}, err`) - this comment also applies to the return statement 5 lines above

Reply:
explicit returns introduced